src/common/tracer.cc

```diff
@@ -55,6 +65,37 @@ bool Tracer::is_enabled() const {
   return g_ceph_context->_conf->jaeger_tracing_enable;
 }
 
+void encode(const jspan_context& span_ctx, bufferlist& bl, uint64_t f) {
+  if (span_ctx.IsValid()) {
+    using namespace opentelemetry;
+    using namespace trace;
+    ENCODE_START(1, 1, bl);
+    encode_nohead(std::string_view((char*)span_ctx.trace_id().Id().data(), TraceId::kSize), bl);
+    encode_nohead(std::string_view((char*)span_ctx.span_id().Id().data(), SpanId::kSize), bl);
+    encode(span_ctx.trace_flags().flags(), bl);
+    ENCODE_FINISH(bl);
+  }
+}
+
+void decode(jspan_context& span_ctx, bufferlist::const_iterator& bl) {
+  if (bl.get_remaining() == 0) {
+    return ;
+  }
+  using namespace opentelemetry;
+  using namespace trace;
+  const char* trace_id_ptr;
+  const char* span_id_ptr;
+  uint8_t flags;
+  DECODE_START(1, bl);
+  bl.get_ptr_and_advance(TraceId::kSize, &trace_id_ptr);
+  bl.get_ptr_and_advance(SpanId::kSize, &span_id_ptr);
```

Comment:
i don't think `get_ptr_and_advance()` is doing what you want here. the bufferlist may have several buffer segments, and there's no guarantee that the `trace_id` and `span_id` are contiguous in memory
```c++
size_t buffer::list::iterator_impl<is_const>::get_ptr_and_advance(
    size_t want, const char **data)
```
you tell the function how many bytes you `want`. if it reaches the end of a buffer segment before that, it returns a value less than `want`

you probably just want to read them into character arrays on the stack:
```c++
std::array<uint8_t, TraceId::kSize> trace_id;
decode(trace_id, bl);
std::array<uint8_t, SpanId::kSize> span_id;
decode(span_id, bl);
```
there's a `nostd::span()` constructor for `std::array&`, so you should be able to pass these arrays directly into `SpanContext`:
```c++
SpanContext ctx(trace_id, span_id, ...
```

Reply:
good point, thanks.

Reply:
done,
just one thing, we can't pass the arrays directly into `SpanContext` because `TraceId` and `SpanId` c'tors are defined as explicit